src/main/scala/org/kframework/minikore/interfaces/Interfaces.scala

```diff
@@ -0,0 +1,493 @@
+
+    def build(_1: CC1, _2: CC2): Pattern
+
+    override def build(contents: Product2[CC1, CC2]): Pattern = build(contents._1, contents._2)
+  }
+
+  object Leaf2 {
+    def unapply[CC1, CC2](arg: Leaf2[CC1, CC2]): Option[(CC1, CC2)] = Some(arg._1, arg._2)
+  }
+
+
+  /**
+    * Node with extra member label, respresenting a node's Label. [[pattern.Application]] that extends this trait.
+    *
+    * @tparam L Type of Label.
+    */
+  sealed trait LabeledNode[L] extends Node with Product1[L] {
+    def build(_1: L, args: Seq[Pattern]): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = build(_1, children)
+  }
+
+  object LabeledNode {
+    def unapply[L](arg: LabeledNode[L]): Option[(L, Seq[Pattern])] = Some(arg._1, arg.args)
+  }
+
+
+  /**
+    * A Node with empty list of Patterns as its args list. [[pattern.Top]], [[pattern.Bottom]] extend this trait.
+    */
+  sealed trait Node0 extends Node {
+    override def args = Seq.empty[Pattern]
+
+    def build(): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.isEmpty)
+      build()
+    }
+  }
+
+  object Node0 {
+    def unapply(arg: Node0): Boolean = true
+  }
+
+
+  /**
+    * A Node with a single pattern in its args list. Extended by [[pattern.Next]], [[pattern.Not]].
+    */
+  sealed trait Node1 extends Node with Product1[Pattern] {
+    override def args = Seq(_1)
+
+    def build(_1: Pattern): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 1)
+      build(children.head)
+    }
+  }
+
+  object Node1 {
+    def unapply(arg: Node1): Option[Pattern] = Some(arg._1)
+  }
+
+
+  /**
+    * A Node with two Patterns in its args list. Extended by [[pattern.Or]], [[pattern.And]], [[pattern.Implies]], [[pattern.Equals]], [[pattern.Rewrite]].
+    */
+  sealed trait Node2 extends Node with Product2[Pattern, Pattern] {
+    def build(_1: Pattern, _2: Pattern): Pattern
+
+    override def args = Seq(_1, _2)
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 2)
+      build(children.head, children(1))
+    }
+  }
+
+  object Node2 {
+    def unapply(arg: Node2): Option[(Pattern, Pattern)] = Some(arg._1, arg._2)
+  }
+
+
+  /**
+    * Extends [[Node2]], and only allows [[pattern.Variable]] as the first element, and [[pattern.Pattern]] in its args list. Extended by [[pattern.Exists]], [[pattern.ForAll]].
+    */
+  sealed trait BinderNode extends Node2 {
+    def build(_1: Variable, _2: Pattern): Pattern
+
+    override val _1: Variable
+
+    override def build(_1: Pattern, _2: Pattern): Pattern = build(_1.asInstanceOf[Variable], _2)
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 2)
+      build(children.head.asInstanceOf[Variable], children(1))
+    }
+  }
+
+  object BinderNode {
+    def unapply(arg: BinderNode): Option[(Variable, Pattern)] = Some(arg._1, arg._2)
+  }
+
+}
+
```

Comment:
As far as I can tell, `Variable` is the only dependency from `object tree` to `object pattern` (other than the type-dependency on `Pattern`).

The only way I can think to remove this dependency is to add a `trait Variable` to the `tree._` object, and have the `pattern.Variable` extend that `tree.Variable`. Or we could just not have a `pattern.Variable` and have `Variable` be defined at the `AST` level (rather than at the `Pattern` level). We would still have a `case class Variable ...` in the implementation.

Thoughts? No immediate benefits of either jump out to me, other than being able to remove a dependency of the `AST` view on the `Pattern` view.